asset/asset.go

```diff
@@ -943,6 +944,447 @@ type GroupKeyReveal interface {
 	GroupPubKey(assetID ID) (*btcec.PublicKey, error)
 }
 
+// NewNonSpendableScriptLeaf creates a new non-spendable tapscript script leaf
+// that includes the specified data. If the data is nil, the leaf will not
+// contain any data but will still be a valid non-spendable script leaf.
+//
+// The script leaf is made non-spendable by including an OP_RETURN opcode at the
+// start of the script. While the script can still be executed, it will always
+// fail and cannot be used to spend funds.
+func NewNonSpendableScriptLeaf(data []byte) lfn.Result[txscript.TapLeaf] {
+	// Construct a script builder and add the OP_RETURN opcode to the start
+	// of the script to ensure that the script is non-executable.
+	scriptBuilder := txscript.NewScriptBuilder().AddOp(txscript.OP_RETURN)
+
+	// Add the data to the script if it is provided.
+	if data != nil {
+		scriptBuilder.AddData(data)
+	}
+
+	// Construct script from the script builder.
+	script, err := scriptBuilder.Script()
+	if err != nil {
+		return lfn.Errf[txscript.TapLeaf]("failed to construct "+
+			"non-spendable script: %w", err)
+	}
+
+	// Create a new tapscript leaf from the script.
+	leaf := txscript.NewBaseTapLeaf(script)
+	return lfn.Ok(leaf)
+}
+
+// TLV types for GroupKeyReveal.
+var (
+	GKRVersion           = new(tlv.TlvType0)
+	GKRInternalKey       = new(tlv.TlvType2)
+	GKRTapsciptRoot      = new(tlv.TlvType4)
+	GKRCustomSubtreeRoot = new(tlv.TlvType7)
+)
+
+func NewGKRVersionRecord(
+	version *uint8) tlv.Record {
```

Comment:
FWIW, for just about all these fields (primitive fields) you can use the new `RecordT` type, here's an example: https://github.com/lightningnetwork/lnd/blob/8ecef03315f749ae8820b4735978ca2e0552de2e/lnwire/musig2.go#L22-L24

So you can just have a single struct, and declare all the fields as attributes, it would look something like: 
```go

type (
   GKRVersionType           = new(tlv.TlvType0)
   GKRInternalKeyType       = new(tlv.TlvType2)
   GKRTapsciptRootType     = new(tlv.TlvType4)
   GKRCustomSubtreeRootType = new(tlv.TlvType7)
)
type GRoupKeyRevealV1 struct {
    Version tlv.RecordT[GKRVersionType, uint32]

    InternalKey tlv.RecordT[GKRInternalKeyType, *btcec.PublicKey]

   TapscriptRoot tlv.RecordT[GKRTapsciptRootType, [sha256.Size]byte]
  
   CustomTreeRoot tlv.Record[GKRCustomSubtreeRootType, [sha256.Size]byte]
}
```

Then your encode+decode is something like: 
```go
func (g *GRoupKeyRevealV1) Encode(w io.Writer) error {
    	stream, err := tlv.NewStream(
		g.Version.Record(), g.InternalKey.Record(), g.TapscriptRoot.Record(),g.CustomTreeRoot.Record(),
	)
	if err != nil {
		return err
	}

	return stream.Encode(w)
}

func (g *GRoupKeyRevealV1) Decode(r io.Reader) error {
  	stream, err := tlv.NewStream(
		g.Version.Record(), g.InternalKey.Record(), g.TapscriptRoot.Record(),g.CustomTreeRoot.Record(),
	)
	if err != nil {
		return err
	}

	return stream.Decode(r)
}

func eGkrV1(w io.Writer, val interface{}, _ *[8]byte) error {
	if v, ok := val.(*GRoupKeyRevealV1); ok {
		return v.Encode(w)
	}

	return tlv.NewTypeForEncodingErr(val, "*GRoupKeyRevealV1")
}

func dGkrV1(r io.Reader, val interface{}, _ *[8]byte, l uint64) error {
	if v, ok := val.(*GRoupKeyRevealV1); ok {
		return v.Decode(r)
	}

	return tlv.NewTypeForDecodingErr(val, "*GRoupKeyRevealV1", l, l)
}

func (g *GRoupKeyRevealV1) Record() tlv.Record {
	sizeFunc := func() uint64 {
		var buf bytes.Buffer
		err := pgEncode(&buf)
		if err != nil {
			panic(err)
		}
		return uint64(len(buf.Bytes()))
	}

	return tlv.MakeDynamicRecord(
		0, p, sizeFunc, eGkrV1, dGkrv1,
	)
}
```

Reply:
Then I think we can even add the `Record()` method to the old interface. So you just call that when makign the stream to handle encode/decode.

Reply:
I played around with using `tlv.RecordT` fields for a while but I couldn't get everything to work correctly.

I got lost in the weeds a bit trying to change all the types and modifying how the fields were used. I got thrown converting `customSubtreeRoot fn.Option[chainhash.Hash]` into a `tlv.OptionalRecordT`. Maybe the key was to drop `chainhash.Hash` in favour of `[sha256.Size]byte`.

I wouldn't mind giving the `tlv.RecordT` approach another go, I've managed to use it in RFQ `requestWireMsgData`, but would you mind if I do that after rc3?